### PR TITLE
Add GKE mode to GCE cloud provider

### PR DESCRIPTION
**Pull request by mwielgus**
Wit this, all autoscaling information will be taken from node pool config, not from the command line.

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 1]
```diff
@@ -88,51 +107,130 @@ func CreateGceManager(configReader io.Reader) (*GceManager, error) {
...
+
+	if mode == ModeGKE {
+		gkeService, err := gke.New(client)
+		if err != nil {
+			return nil, err
+		}
+		manager.gkeService = gkeService
+		err = manager.fetchAllNodePools()
+		if err != nil {
+			glog.Errorf("Failed to fech node pools: %v", err)
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: s/fech/fetch

Comment (MaciekPytel) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Why do we ignore this error? Shouldn't we just consider that a failure to build manager and return the error?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 2]
```diff
@@ -88,51 +107,130 @@ func CreateGceManager(configReader io.Reader) (*GceManager, error) {
...
-	defer m.cacheMutex.Unlock()
+func (m *GceManager) assertGKE() {
+	if m.mode != ModeGKE {
+		panic(fmt.Errorf("This should run only in GKE mode"))
+	}
+}
+
+// Gets all registered node pools
+func (m *GceManager) fetchAllNodePools() error {
+	m.assertGKE()
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Why not just return error if it's not GKE? Or, if we really, really want to guarantee we kill the process if this happens, use Fatalf as we do elsewhere?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 3]
```diff
@@ -88,51 +107,130 @@ func CreateGceManager(configReader io.Reader) (*GceManager, error) {
...
+			if autoscaled {
+				mig.minSize = int(nodePool.Autoscaling.MinNodeCount)
+				mig.maxSize = int(nodePool.Autoscaling.MaxNodeCount)
+			} else if autoprovisioned {
+				mig.minSize = minAutoprovisionedSize
+				mig.maxSize = maxAutoprovisionedSize
+			}
+			m.RegisterMig(mig)
+		}
+		// TODO - unregister migs
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: When are you planning to implement this? Sounds pretty important, especially given autoprovisioning.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 4]
```diff
@@ -88,51 +107,130 @@ func CreateGceManager(configReader io.Reader) (*GceManager, error) {
...
+func (m *GceManager) fetchAllNodePools() error {
+	m.assertGKE()
+
+	nodePoolsResponse, err := m.gkeService.Projects.Zones.Clusters.NodePools.List(m.projectId, m.zone, m.clusterName).Do()
+	if err != nil {
+		return err
+	}
+	for _, nodePool := range nodePoolsResponse.NodePools {
+		autoprovisioned := strings.Contains("name", nodeAutoprovisioningPrefix)
+		autoscaled := nodePool.Autoscaling != nil && nodePool.Autoscaling.Enabled
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Is autoprovisioned and autoscaled meant to be mutually exclusive? Or is the semantic different if it's both autoprovisioned and autoscaled?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 2 above)
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 3 above)
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Yes. Next PR.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 4 above)
Comment (MaciekPytel) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: I mean I can figure out from code what will happen, my question is whether this is what we want to happen?

Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Yes. Autoprovisioned doesn't have min/max.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go [hunk 5]
```diff
@@ -88,51 +107,130 @@ func CreateGceManager(configReader io.Reader) (*GceManager, error) {
...
+		}
+	}
 
-	m.migs = append(m.migs, &migInformation{
-		config: mig,
-	})
+	if !updated {
+		glog.V(1).Infof("Registering %s/%s/%s", mig.GceRef.Project, mig.GceRef.Zone, mig.GceRef.Name)
+		m.migs = append(m.migs, &migInformation{
+			config: mig,
```
Comment (MaciekPytel) [inline cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Why only populate one field in this struct? It seems basename is pretty important

**Comment by MaciekPytel** [comment]
My biggest problem is how node groups are meant to be configured in GKE? With this PR it seem it would still parse `--nodes` flag, but it would also get the same information from fetchAllNodePools(). What if they conflict? It seems one would silently override the other? I think we need to explicitly decide which one is going to be our source of truth.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 4 above)
Comment (MaciekPytel) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: What if someone runs gcloud and updates autoscaling for autoprovisioned node pool?

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 1 above)
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: Done.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 4 above)
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: We won't allow that.

File: cluster-autoscaler/cloudprovider/gce/gce_manager.go (hunk 5 above)
Comment (mwielgus) [reply cluster-autoscaler/cloudprovider/gce/gce_manager.go]: It will be populated a moment later in regenerate cache.

**Comment by mwielgus** [comment]
We will not use --nodes for GKE.

**Comment by MaciekPytel** [comment]
Ok, so disabling --nodes for GKE is one of the things added in follow-up PR? Maybe it would be worth creating issue to track this stuff, so we don't forget? Or just add comment in #225 after each PR listing carried over TODOs?

**Pull request merged**